espnet/nets/chainer_backend/e2e_asr_transformer.py

```diff
@@ -169,80 +155,38 @@ def forward(self, xs, ilens, ys_pad, calculate_attentions=False):
             chainer.Variable (Optional): Output of the encoder.
 
         """
-        xp = self.xp
-        with chainer.no_backprop_mode():
-            xs = xp.array(xs)
-            eos = np.array([self.eos], 'i')
-            sos = np.array([self.sos], 'i')
-            ys_out = [F.concat([y, eos], axis=0) for y in ys_pad]
-            ys = [F.concat([sos, y], axis=0) for y in ys_pad]
-            ys = F.pad_sequence(ys, padding=self.eos)
-            ys_out = F.pad_sequence(ys_out, padding=-1)
-        ys = xp.array(ys.data)
-        ys_out = chainer.Variable(xp.array(ys_out.data))
-        ys_pad_cpu = [y.astype(np.int32) for y in ys_pad]
-
-        logging.info(self.__class__.__name__ + ' input lengths: ' + str(ilens))
-        # Encode Sources
-        # xs: utt x frame x dim
-        logging.debug('Init size: ' + str(xs.shape))
-        logging.debug('Out size: ' + str(ys.shape))
-        # Dims along enconder and decoder: batchsize * length x dims
-        xs, x_mask, ilens = self.encoder(xs, ilens)
-        logging.info(self.__class__.__name__ + ' input lengths: ' + str(ilens))
-        logging.info(self.__class__.__name__ + ' output lengths: ' + str(xp.array([y.shape[0] for y in ys_out])))
-        xy_mask = self.make_attention_mask(ys, xp.array(x_mask))
-        yy_mask = self.make_attention_mask(ys, ys)
-        yy_mask *= self.make_history_mask(ys)
-        batch, length = ys.shape
-        ys = self.decoder(ys, yy_mask, xs, xy_mask)
-        if calculate_attentions:
-            return xs
+        alpha = self.mtlalpha
 
-        # Compute Attention Loss
-        loss_att = self.criterion(ys, ys_out, batch, length)
-        acc = F.accuracy(ys, ys_out.reshape((-1)).data, ignore_label=self.ignore_id)
+        # 1. Encoder
+        xs, x_mask, ilens = self.encoder(xs, ilens)
 
-        # Compute CTC Loss and CER CTC
+        # 2. CTC loss
         cer_ctc = None
-
-        if self.ctc is None:
+        if alpha == 0.0:
             loss_ctc = None
         else:
-            xs = xs.reshape(batch, -1, self.dims)
-            xs = [xs[i, :ilens[i], :] for i in range(len(ilens))]
-            loss_ctc = self.ctc(xs, ys_pad_cpu)
+            _ys = [y.astype(np.int32) for y in ys_pad]
+            loss_ctc = self.ctc.forward_from_transformer(xs, _ys)
             if self.error_calculator is not None:
                 with chainer.no_backprop_mode():
                     ys_hat = chainer.backends.cuda.to_cpu(self.ctc.argmax(xs).data)
-                cer_ctc = self.error_calculator(ys_hat, ys_pad_cpu, is_ctc=True)
+                cer_ctc = self.error_calculator(ys_hat, ys_pad, is_ctc=True)
 
-        # Compute cer/wer
-        with chainer.no_backprop_mode():
-            y_hats = ys.reshape((batch, length, -1))
-            y_hats = chainer.backends.cuda.to_cpu(F.argmax(y_hats, axis=2).data)
+        # 3. Decoder
+        ys = self.decoder(ys_pad, xs, x_mask)
+        if calculate_attentions:
```

Comment:
instead of return nothing, I suggest to create new method like `def calculate_attentions(self, xs, xs_mask, ys_pad)`